File: csharp/lib/Property/ColorSpaceProperty.cs
/*
 * SPDX-FileCopyrightText: 2017-2022 Sony Semiconductor Solutions Corporation
 *
 * SPDX-License-Identifier: Apache-2.0
 */

using MessagePack;

namespace SensCord
{
    /// <summary>
    /// Property of color space type for YUV.
    /// </summary>
    [MessagePackObject]
    public class ColorSpaceProperty : IBaseProperty
    {
        /// <summary>
        /// Property key.
        /// </summary>
        public static readonly string ConstKey = "color_space_property";

        /// <summary>
        /// Property key.
        /// </summary>
        [IgnoreMember]
        public string Key { get; } = ConstKey;

        /// <summary>
        /// Encoding.
        /// </summary>
        [Key("encoding")]
        public YCbCrEncoding Encoding { get; set; }

        /// <summary>
        /// Quantization.
        /// </summary>
        [Key("quantization")]
        public YCbCrQuantization Quantization { get; set; }

        /// <summary>
        /// To string.
        /// </summary>
        /// <returns>A string that represents the current object.</returns>
        public override string ToString()
        {
            return $"{this.Key}(encoding={this.Encoding}, quantization={this.Quantization})";
        }
    }

    /// <summary>
    /// Encoding types for YUV (YCbCr).
    /// </summary>
    public enum YCbCrEncoding
    {
        /// <summary>
        /// Undefined.
        /// </summary>
        Undefined,

        /// <summary>
        /// BT601.
        /// </summary>
        BT601,

        /// <summary>
        /// BT709.
        /// </summary>
        BT709,

        /// <summary>
        /// BT2020.
        /// </summary>
        BT2020,

        /// <summary>
        /// BT2100.
        /// </summary>
        BT2100,
    }

    /// <summary>
    /// Quantization types for YUV (YCbCr).
    /// </summary>
    public enum YCbCrQuantization
    {
        /// <summary>
        /// BT2020.
        /// </summary>
        Undefined,

        /// <summary>
        /// Full range (Y: 0-255,  C: 0-255).
        /// </summary>
        FullRange,

        /// <summary>
        /// Limited range (Y: 16-235, C: 16-240).
        /// </summary>
        LimitedRange,

        /// <summary>
        /// Super white.
        /// </summary>
        SuperWhite,
    }
}

File: java/senscord/src/senscord_properties/ColorSpaceProperty.java
/*
 * SPDX-FileCopyrightText: 2019-2022 Sony Semiconductor Solutions Corporation
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package senscord_properties;

import com.sun.jna.Structure;

import java.util.Arrays;
import java.util.List;

/**
 * Property of color space type for YUV.
 * 
 * @see "senscord::ColorSpaceProperty"
 */
public class ColorSpaceProperty extends Structure {
    public int encoding;
    public int quantization;

    public static class ByReference extends ColorSpaceProperty implements Structure.ByReference {
    }

    @Override
    protected List<String> getFieldOrder() {
        return Arrays.asList("encoding", "quantization");
    }

    /**
     * Encoding types for YUV (YCbCr).
     * 
     * @see "senscord::YCbCrEncoding"
     */
    public static interface Encoding {
        public static final int SENSCORD_YCBCR_ENCODING_UNDEFINED = 0;
        public static final int SENSCORD_YCBCR_ENCODING_BT601 = 1;
        public static final int SENSCORD_YCBCR_ENCODING_BT709 = 2;
        public static final int SENSCORD_YCBCR_ENCODING_BT2020 = 3;
        public static final int SENSCORD_YCBCR_ENCODING_BT2100 = 4;
    }

    /**
     * Quantization types for YUV (YCbCr).
     * 
     * @see "senscord::YCbCrQuantization"
     */
    public static interface Quantization {
        public static final int SENSCORD_YCBCR_QUANTIZATION_UNDEFINED = 0;

        /**
         * Y: 0 - 255, C: 0 - 255.
         */
        public static final int SENSCORD_YCBCR_QUANTIZATION_FULL_RANGE = 1;

        /**
         * Y: 16 - 255, C: 16 - 240.
         */
        public static final int SENSCORD_YCBCR_QUANTIZATION_KIMITED_RANGE = 2;
        public static final int SENSCORD_YCBCR_QUANTIZATION_SUPER_WHITE = 3;
    }

    /**
     * Get color space encoding.
     * 
     * @return color space encoding
     */
    public int GetEncoding() {
        return encoding;
    }

    /**
     * Get color space quantization.
     * 
     * @return color space encoding
     */
    public int GetQuantization() {
        return quantization;
    }

    /**
     * Set color space encoding.
     * 
     * @param encoding color space encoding. {@link Encoding}
     */
    public void SetEncoding(int encoding) {
        this.encoding = encoding;
    }

    /**
     * Set color space quantization.
     * 
     * @param quantization color space quantization {@link Quantization}}
     */
    public void SetQuantization(int quantization) {
        this.quantization = quantization;
    }
}
